### PR TITLE
Dont update state delete state when undelete is disabled

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
When undelete is disabled, DELETE will be the final state of the blob's life cycle. 
In this case, when we iterate through IndexSegment in findEntriesSince, we have to call update state for the DELETE record, since we already know it's the final state.

This PR should greatly improve the performance of replicating through a log segment that is full of DELETEs.

### CODE REVIEW

**Review by SophieGuo410 — APPROVED** [review]

**Comment by snalli** [comment]
some unittest for this would be good.

**Comment by justinlin-linkedin** [comment]
> some unittest for this would be good.

Added more test cases in IndexTest

File: ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java [hunk 1]
```diff
@@ -225,6 +227,13 @@ public int compare(IndexEntry e1, IndexEntry e2) {
     this.incarnationId = incarnationId;
     this.maxInMemoryIndexSizeInBytes = config.storeIndexMaxMemorySizeBytes;
     this.maxInMemoryNumElements = config.storeIndexMaxNumberOfInmemElements;
+    if (config.storeCompactionFilter
```
Comment (ankagrawal) [inline ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java:230]: .equals() for string comparison

File: ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java [hunk 2]
```diff
@@ -1947,8 +1956,11 @@ private void updateStateForMessages(List<MessageInfo> messageEntries) throws Sto
     ListIterator<MessageInfo> messageEntriesIterator = messageEntries.listIterator();
     while (messageEntriesIterator.hasNext()) {
       MessageInfo messageInfo = messageEntriesIterator.next();
-      // for all the message info, we use most recent ref even for delete. since a deleted entry can be undeleted.
-      // ok to use most recent ref
+      // If the message is already deleted and DELETE is the final state of the blob, then we don't have
+      // to update the state for this message.
+      if (isDeleteFinalStateOfBlob && messageInfo.isDeleted()) {
+        continue;
```
Comment (ankagrawal) [inline ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java:1962]: Should we add unit tests for this?

File: ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java (hunk 1 above)
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java]: updated.

File: ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java (hunk 2 above)
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java:1962]: We already have findEntriesSinceTest in the IndexTest file. I added a new parameter to that unit test class for this.

**Review by ankagrawal — APPROVED** [review]

**Pull request merged**